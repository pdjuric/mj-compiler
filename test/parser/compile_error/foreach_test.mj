program foreach_test

class A {
    int ident;
    int designator[];
    {
        A() {
            this.designator = new int[6];
            designator[0] = 0;
            designator[1] = 1;
            designator[2] = 2;
            designator[3] = 3;
            designator[4] = 4;
            designator[5] = 5;
        }
    }
}

class B extends A{
    B arr[];
    {
        B() {}

        B(boolean t) { 
            this.arr = new B[2];
            arr[0] = new B();
            arr[1] = new B();
        }
    }
}

int g_i;
int g_arr[];

{
  
    void main() int cnt[]; A a_a, t; B b_b; int l_i; char l_c; {
        a_a = new A();
        b_b = new B();
        g_arr = new int[3];

        a_a.desingator.foreach(l_i => {
            l_i++;
            l_i = 2;
        });

        a_a.desingator.foreach(a_a.ident =>  print('x'); );

        a_a.desingator.foreach(g_i => {
            print('o', 1); print('k', 1); print(eol);
        });

        g_i.foreach(l_i => print('a'););        //designator nije array

        l_i.foreach(l_c => print('b'););

        a_a.foreach(l_i => print('c'););

        b_b.arr.foreach( t =>
            b_b.arr.foreach(l_i => t = 5; );
        );

    }
}