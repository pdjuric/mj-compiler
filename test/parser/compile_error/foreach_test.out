INFO  16:59:31,003 - line  40 - Access to global variable g_arr [Var g_arr: Arr of int, 1, 0 ].
INFO  16:59:31,003 - line  42 - Access to local variable a_a [Var a_a: Class A, 1, 1 ].
INFO  16:59:31,004 - line  42 - Access to member designator of class A [Fld designator: Arr of int, 2, 1 ].
INFO  16:59:31,004 - line  43 - Access to local variable l_i [Var l_i: int, 5, 1 ].
ERROR 16:59:31,005 - line  43 - Value of ForEach loop variable cannot be changed.
INFO  16:59:31,005 - line  44 - Access to local variable l_i [Var l_i: int, 5, 1 ].
ERROR 16:59:31,005 - line  44 - Value of ForEach loop variable cannot be changed.
INFO  16:59:31,006 - line  49 - Access to local variable a_a [Var a_a: Class A, 1, 1 ].
INFO  16:59:31,006 - line  49 - Access to member designator of class A [Fld designator: Arr of int, 2, 1 ].
INFO  16:59:31,006 - line  50 - Call to global function print.
INFO  16:59:31,007 - line  50 - Call to global function print.
INFO  16:59:31,007 - line  50 - Access to constant eol [Con eol: char, 10, 0 ].
INFO  16:59:31,007 - line  50 - Call to global function print.
INFO  16:59:31,007 - line  53 - Access to global variable g_i [Var g_i: int, 0, 0 ].
ERROR 16:59:31,008 - line  53 - ForEach loop invoked on non-array variable.
INFO  16:59:31,008 - line  53 - Call to global function print.
INFO  16:59:31,008 - line  55 - Access to local variable l_i [Var l_i: int, 5, 1 ].
ERROR 16:59:31,008 - line  55 - ForEach loop invoked on non-array variable.
INFO  16:59:31,008 - line  55 - Call to global function print.
INFO  16:59:31,010 - line  57 - Access to local variable a_a [Var a_a: Class A, 1, 1 ].
ERROR 16:59:31,010 - line  57 - ForEach loop invoked on non-array variable.
INFO  16:59:31,010 - line  57 - Call to global function print.
INFO  16:59:31,011 - line  59 - Access to local variable b_b [Var b_b: Class B, 4, 1 ].
INFO  16:59:31,011 - line  59 - Access to member arr of class B [Fld arr: Arr of Class, 3, 1 ].
INFO  16:59:31,011 - line  60 - Access to local variable b_b [Var b_b: Class B, 4, 1 ].
INFO  16:59:31,011 - line  60 - Access to member arr of class B [Fld arr: Arr of Class, 3, 1 ].
ERROR 16:59:31,011 - line  60 - Type of ForEach loop variable l_i does not match the type of array elements.
INFO  16:59:31,012 - line  60 - Access to local variable t [Var t: Class A, 2, 1 ].
ERROR 16:59:31,012 - line  60 - Operands are not compatible for assignment.
ERROR 16:59:31,012 - line  60 - Value of ForEach loop variable cannot be changed.
INFO  16:59:31,012 - line  63 - Access to local variable b_b [Var b_b: Class B, 4, 1 ].
INFO  16:59:31,013 - line  63 - Access to member arr of class B [Fld arr: Arr of Class, 3, 1 ].
INFO  16:59:31,013 - line  64 - Access to local variable b_b [Var b_b: Class B, 4, 1 ].
INFO  16:59:31,013 - line  64 - Access to member arr of class B [Fld arr: Arr of Class, 3, 1 ].
INFO  16:59:31,013 - line  64 - Access to local variable t [Var t: Class A, 2, 1 ].
ERROR 16:59:31,013 - line  64 - Operands are not compatible for assignment.
ERROR 16:59:31,013 - line  64 - Value of ForEach loop variable cannot be changed.